p2p/protocol/identify/peer_loop.go

```diff
@@ -52,23 +55,31 @@ func newPeerHandler(pid peer.ID, ids *IDService) *peerHandler {
 
 		pushCh:  make(chan struct{}, 1),
 		deltaCh: make(chan struct{}, 1),
+
+		// defaults to false
+		isRunningAtomic: abool.New(),
 	}
 
 	return ph
 }
 
+// should be idempotent.
 func (ph *peerHandler) start() {
-	ctx, cancel := context.WithCancel(context.Background())
-	ph.ctx = ctx
-	ph.cancel = cancel
+	if ph.isRunningAtomic.SetToIf(false, true) {
+		ctx, cancel := context.WithCancel(context.Background())
+		ph.ctx = ctx
+		ph.cancel = cancel
 
-	ph.wg.Add(1)
-	go ph.loop()
+		ph.wg.Add(1)
+		go ph.loop()
+	}
 }
 
 func (ph *peerHandler) close() error {
-	ph.cancel()
-	ph.wg.Wait()
+	if ph.isRunningAtomic.SetToIf(true, false) {
+		ph.cancel()
+		ph.wg.Wait()
+	}
 	return nil
 }
```

Comment:
We still need to _wait_ before writing to the "closed" channel. Otherwise, the second close will skip the wait.